### DISCUSSION TITLE
oauth2 sign-up

### DISCUSSION THREAD

**Discussion by cheetahbyte** [pocketbase/pocketbase#985] [Q&A] [answered]
Hey there,
I'm having some issues implementing an OAuth2 sign-up. I'm using the newest version (`v0.8.0-rc1`). 
My problem is simply that I don't really know what to put in the `createData` parameter.
Hopefully someone can give me a hint :sweat_smile: 
Thanks!

**Comment by ganigeorgiev** [comment, marked as answer]
`createData` is optional and it is there to allow you to set the value of your custom user fields during account creation (eg. `role`). If your `users` collection doesn't have custom required fields, you can skip it.

The documentation will be updated soon, but if you are looking for a OAuth2 guide you could check https://pocketbase.io/docs/manage-users/#auth-via-oauth2. For v0.8.0-rc you'll need to only replace `client.users.*` with `client.collection("users").*` and rename `authViaOAuth2`  with `authWithOAuth2`:

```js
// index.html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>OAuth2 links page</title>
    <script src="https://code.jquery.com/jquery-3.6.0.slim.min.js"></script>
</head>
<body>
    <ul id="list">
        <li>Loading OAuth2 providers...</li>
    </ul>

    <script src="https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@rc/dist/pocketbase.umd.js"></script>
    <script type="text/javascript">
        const pb = new PocketBase('http://127.0.0.1:8090');
        const redirectUrl = 'http://127.0.0.1:8090/redirect.html';

        async function loadLinks() {
            const authMethods = await pb.collection('users').listAuthMethods();
            const listItems = [];

            for (const provider of authMethods.authProviders) {
                const $li = $(`<li><a>Login with ${provider.name}</a></li>`);

                $li.find('a')
                    .attr('href', provider.authUrl + redirectUrl)
                    .data('provider', provider)
                    .click(function () {
                        // store provider's data on click for verification in the redirect page
                        localStorage.setItem('provider', JSON.stringify($(this).data('provider')));
                    });

                listItems.push($li);
            }

            $('#list').html(listItems.length ? listItems : '<li>No OAuth2 providers.</li>');
        }

        loadLinks();
    </script>
</body>
</html>


// redirect.html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>OAuth2 redirect page</title>
</head>
<body>
    <pre id="content">Authenticating...</pre>

    <script src="https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@rc/dist/pocketbase.umd.js"></script>
    <script type="text/javascript">
        const pb = new PocketBase("http://127.0.0.1:8090");
        const redirectUrl = 'http://127.0.0.1:8090/redirect.html';

        // parse the query parameters from the redirected url
        const params = (new URL(window.location)).searchParams;

        // load the previously stored provider's data
        const provider = JSON.parse(localStorage.getItem('provider'))

        // compare the redirect's state param and the stored provider's one
        if (provider.state !== params.get('state')) {
            throw "State parameters don't match.";
        }

        // authenticate
        pb.collection('users').authWithOAuth2(
            provider.name,
            params.get('code'),
            provider.codeVerifier,
            redirectUrl,
            // pass optional user create data
            {
                emailVisibility: false,
            }
        ).then((authData) => {
            document.getElementById('content').innerText = JSON.stringify(authData, null, 2);
        }).catch((err) => {
            document.getElementById('content').innerText = "Failed to exchange code.\n" + err;
        });
    </script>
</body>
</html>
```

**Reply by cheetahbyte** [reply]
Hey @ganigeorgiev,
Thank you for your detailed answer. I always get a 400 "Failed to authenticate" when I try to do this. Do you know what the reason is? Both SDK and server are up to date.

**Reply by ganigeorgiev** [reply]
@cheetahbyte A lot of things could result to a 400 error - invalid code, incorrect verifier challenge, incorrectly setup redirect url, etc. Try running the PocketBase executable with the `--debug` flag and/or check the error logs in the "Admin UI > Logs" for a more detailed error message.

**Reply by cheetahbyte** [reply]
Here is the request log
![image](https://user-images.githubusercontent.com/40123243/201153112-68ebab86-1d81-41d2-9d7f-f89265b8d70d.png)

**Reply by ganigeorgiev** [reply]
@cheetahbyte It seems that you are missing something when sending the request. Could you provide a code sample how you call `authWithOAuth2`?

**Reply by cheetahbyte** [reply]
@ganigeorgiev 
( i'm using svelte btw)
```ts
<script lang="ts">

onMount(async() => {
    await client.collection("users").authWithOAuth2(provider.name, code, provider.codeChallenge, "http://localhost:3000")
})
</script>
```
i'm getting the provider object from `client.collection("users").listAuthMethods`

**Reply by ganigeorgiev** [reply]
@cheetahbyte Make sure that the arguments you are passing (eg. `code`, `provider.*`, etc.) are not empty. Also, check if `redirectUrl` matches the one you've set when you created the "Login with ..." links.

To rule out issue with the OAuth2 integration, you can add the above 2 files from the example in your `pb_public` directory and try to test the integration with them (you'll need to also adjust the redirect url in you OAuth2 app).

**Reply by cheetahbyte** [reply]
Such a stupid mistake on my part. I had a typo in the redirect url :|

Thanks for your help!

**Reply by ganigeorgiev** [reply]
Also note that you have to pass the `provider.codeVerifier` not `provider.codeChallenge`.

**Comment by dschissler** [comment]
@ganigeorgiev That is a great example.  I just searched here because I couldn't find anything at https://pocketbase.io/docs/api-authentication.  It's unclear which page that this content should live in due to the admin panel usage.

Perhaps you should move the "Auth via OAuth2" content to the `api-authentication` page?  On the page where it is now you could do it like this:

```javascript
import PocketBase from 'pocketbase';

const client = new PocketBase('http://127.0.0.1:8090');

...

const authMethods = await client.users.listAuthMethods();
// Display the authorization methods as links
...
```

Then below this you could link to `api-authorization` page for more details.  This allows you to discuss the admin page and to give a small taste of what it looks like in code.  Otherwise it's not great for discovering it.  For example, I went through the entire documentation in order, but I wasn't ready for the in-depth code example yet.  Then I couldn't find it where it seemed like it ought to be.

**Reply by dschissler** [reply]
Also, when someone is reading it for the first time they'll probably click into the more detailed `api-authentication` page in a tab, look at it briefly and then retreat backwards to continue reading the documentation for the larger idea.  I see the "Manage Users" section as mostly being for someone who hasn't yet even played with the admin panel.  It's not where they'll look later when they're at the stage that I am now.

**Reply by ganigeorgiev** [reply]
@dschissler This is already changed in the v0.8.0 docs. The OAuth2 web example has been moved under the Authentication section. It will be publicly available with the stable v0.8.0 release soon. It is also linked in the "API preview" section in the Admin UI:

![oauth2_screenshot](https://i.imgur.com/WezuJzM.png)

Also please note that the OAuth2 flow will be simplified in the near future per #55.

**Reply by rajesh6161** [reply]
@cheetahbyte if you are looking for React oAuth then I have made a repo regarding the same please have a look:
https://github.com/rajesh6161/pocketbase-oauth-demo
